File: crates/kuben-core/src/policy.rs
//! Environment policy and deployment approvals (M4.1; plan §8.2, §13).
//!
//! An environment's policy is an immutable revision: who may start
//! deployments there, how many distinct people must approve a change
//! before it is delivered, who may approve, and how long a request waits.
//! Protection is policy, never inferred from a name. The rules here are
//! pure; the store applies them inside the transaction that locks the run.
//!
//! ```text
//! planned ──(approvals required)──► awaitingApproval ──(enough approvals)──► pendingDelivery
//!                                        │ reject / expiry
//!                                        ▼
//!                                    cancelled
//! ```

use serde::{Deserialize, Serialize};

use crate::perm::{Perm, Role};

/// Most approvals a policy may require.
pub const MAX_APPROVALS: u8 = 5;
/// Shortest and longest time a change may wait for its approvals.
pub const MIN_APPROVAL_TTL_SECS: u32 = 5 * 60;
pub const MAX_APPROVAL_TTL_SECS: u32 = 30 * 24 * 3600;
/// Default wait for approvals: one week.
pub const DEFAULT_APPROVAL_TTL_SECS: u32 = 7 * 24 * 3600;
/// Longest approval comment kept.
pub const MAX_COMMENT_CHARS: usize = 1024;

/// Why a policy was refused.
#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
pub enum PolicyError {
    #[error("at most {MAX_APPROVALS} approvals can be required")]
    TooManyApprovals,
    #[error("the deploy role `{0}` cannot deploy")]
    DeployRole(Role),
    #[error("the approve role `{0}` cannot approve releases")]
    ApproveRole(Role),
    #[error("approvals must wait between {MIN_APPROVAL_TTL_SECS} and {MAX_APPROVAL_TTL_SECS} seconds")]
    ApprovalTtl,
}

/// One revision of an environment's policy.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnvironmentPolicy {
    /// Distinct approvers a change needs before delivery; 0 for none.
    pub required_approvals: u8,
    /// The weakest role that may start deployments here.
    pub deploy_role: Role,
    /// The weakest role that may approve deployments here.
    pub approve_role: Role,
    /// How long a change waits for its approvals before it is cancelled.
    pub approval_ttl_secs: u32,
}

/// What a deployment run changes, as far as approval is concerned.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ChangeKind {
    Deploy,
    Rollback,
    Promotion,
    /// An automatic deploy of a verified build.
    Build,
    /// The same release and configuration, pods replaced.
    Restart,
    /// The same release and configuration, delivered by the agent.
    Handover,
    /// The same release and configuration with a secret's new revision: new
    /// values reach production like any other change.
    Rotation,
}

impl EnvironmentPolicy {
    /// No approvals; developers deploy.
    #[must_use]
    pub const fn open() -> Self {
        Self {
            required_approvals: 0,
            deploy_role: Role::Developer,
            approve_role: Role::Admin,
            approval_ttl_secs: DEFAULT_APPROVAL_TTL_SECS,
        }
    }

    /// One approval by an admin who did not ask for the change.
    #[must_use]
    pub const fn production() -> Self {
        Self {
            required_approvals: 1,
            ..Self::open()
        }
    }

    /// The first policy of a new environment.
    #[must_use]
    pub const fn initial(production: bool) -> Self {
        if production {
            Self::production()
        } else {
            Self::open()
        }
    }

    pub fn validate(&self) -> Result<(), PolicyError> {
        if self.required_approvals > MAX_APPROVALS {
            return Err(PolicyError::TooManyApprovals);
        }
        if !self.deploy_role.grants(Perm::AppDeploy) {
            return Err(PolicyError::DeployRole(self.deploy_role));
        }
        if !self.approve_role.grants(Perm::ReleaseApprove) {
            return Err(PolicyError::ApproveRole(self.approve_role));
        }
        if !(MIN_APPROVAL_TTL_SECS..=MAX_APPROVAL_TTL_SECS).contains(&self.approval_ttl_secs) {
            return Err(PolicyError::ApprovalTtl);
        }
        Ok(())
    }

    /// Approvals a change of `kind` needs. Restarts and handovers change
    /// neither the release nor the configuration.
    #[must_use]
    pub const fn approvals_for(&self, kind: ChangeKind) -> u8 {
        match kind {
            ChangeKind::Restart | ChangeKind::Handover => 0,
            _ => self.required_approvals,
        }
    }

    /// Whether a caller whose strongest role on the environment is `role`
    /// may start a deployment there.
    #[must_use]
    pub fn may_deploy(&self, role: Option<Role>) -> bool {
        role.is_some_and(|r| r.rank() >= self.deploy_role.rank() && r.grants(Perm::AppDeploy))
    }

    /// Whether a caller whose strongest role on the environment is `role`
    /// may approve there.
    #[must_use]
    pub fn may_approve(&self, role: Option<Role>) -> bool {
        role.is_some_and(|r| r.rank() >= self.approve_role.rank() && r.grants(Perm::ReleaseApprove))
    }

    /// Whether `self` protects less than `current` in any respect. Weakening
    /// production protection is an owner's decision.
    #[must_use]
    pub const fn weakens(&self, current: &Self) -> bool {
        self.required_approvals < current.required_approvals
            || self.deploy_role.rank() < current.deploy_role.rank()
            || self.approve_role.rank() < current.approve_role.rank()
            || self.approval_ttl_secs > current.approval_ttl_secs
    }
}

impl Default for EnvironmentPolicy {
    fn default() -> Self {
        Self::open()
    }
}

/// An approver's answer.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Decision {
    Approve,
    Reject,
}

impl Decision {
    /// The stored name of the decision.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Approve => "approved",
            Self::Reject => "rejected",
        }
    }
}

/// Why a decision was refused. Nothing is recorded for a refusal.
#[derive(Clone, Copy, Debug, PartialEq, Eq, thiserror::Error)]
pub enum ApprovalError {
    #[error("the deployment is not waiting for approval")]
    NotAwaiting,
    #[error("the approval window of the deployment has closed")]
    Expired,
    #[error("whoever asked for a deployment cannot approve it")]
    SelfApproval,
    #[error("this approver has already decided on the deployment")]
    AlreadyDecided,
    #[error("the deployment changed since it was shown: reload and decide again")]
    StalePlan,
}

/// What a run waiting for approval looks like to a decision.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Pending<'a> {
    /// `kind:principal` of whoever asked for the run.
    pub requested_by: &'a str,
    pub required: u8,
    /// Distinct approvals recorded so far.
    pub approved: u8,
    pub expires_at: i64,
    pub plan_hash: &'a [u8],
}

/// The outcome of a valid decision.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Tally {
    /// More approvals are needed.
    Waiting { remaining: u8 },
    /// Enough approvals: the run may be delivered.
    Approved,
    /// One rejection cancels the run.
    Rejected,
}

/// The principal part of `kind:principal`.
#[must_use]
pub fn principal(actor: &str) -> &str {
    actor.split_once(':').map_or(actor, |(_, id)| id)
}

/// Decide whether `approver` (a user id) may record `decision` on `run` at
/// `now`, having seen `plan_hash`, and what the run becomes.
///
/// # Errors
///
/// The [`ApprovalError`] that refuses the decision.
pub fn decide(
    run: &Pending<'_>,
    approver: &str,
    decided_before: bool,
    decision: Decision,
    plan_hash: &[u8],
    now: i64,
) -> Result<Tally, ApprovalError> {
    if now >= run.expires_at {
        return Err(ApprovalError::Expired);
    }
    // A requester may withdraw nothing and approve nothing: only others decide.
    if principal(run.requested_by) == approver {
        return Err(ApprovalError::SelfApproval);
    }
    if decided_before {
        return Err(ApprovalError::AlreadyDecided);
    }
    if plan_hash != run.plan_hash {
        return Err(ApprovalError::StalePlan);
    }
    Ok(match decision {
        Decision::Reject => Tally::Rejected,
        Decision::Approve => {
            let count = run.approved.saturating_add(1);
            if count >= run.required {
                Tally::Approved
            } else {
                Tally::Waiting {
                    remaining: run.required - count,
                }
            }
        }
    })
}

/// Hex of a plan hash, as the API shows it.
#[must_use]
pub fn hex(bytes: &[u8]) -> String {
    use std::fmt::Write as _;
    bytes
        .iter()
        .fold(String::with_capacity(bytes.len() * 2), |mut s, b| {
            let _ = write!(s, "{b:02x}");
            s
        })
}

/// A plan hash given as hex; `None` when it is not hex.
#[must_use]
pub fn unhex(text: &str) -> Option<Vec<u8>> {
    if !text.len().is_multiple_of(2) || text.len() > 128 {
        return None;
    }
    (0..text.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(text.get(i..i + 2)?, 16).ok())
        .collect()
}

#[cfg(test)]
mod tests {
    use proptest::prelude::*;

    use super::*;

    const HASH: &[u8] = &[1, 2, 3];

    fn pending(required: u8, approved: u8) -> Pending<'static> {
        Pending {
            requested_by: "user:alice",
            required,
            approved,
            expires_at: 1_000,
            plan_hash: HASH,
        }
    }

    #[test]
    fn production_needs_one_admin_approval() {
        let p = EnvironmentPolicy::initial(true);
        assert_eq!(p.required_approvals, 1);
        assert_eq!(p.approve_role, Role::Admin);
        assert_eq!(EnvironmentPolicy::initial(false).required_approvals, 0);
        p.validate().expect("valid");
        EnvironmentPolicy::default().validate().expect("valid");
    }

    #[test]
    fn invalid_policies_are_refused() {
        let base = EnvironmentPolicy::production();
        let cases = [
            (
                EnvironmentPolicy {
                    required_approvals: 6,
                    ..base
                },
                PolicyError::TooManyApprovals,
            ),
            (
                EnvironmentPolicy {
                    deploy_role: Role::Viewer,
                    ..base
                },
                PolicyError::DeployRole(Role::Viewer),
            ),
            (
                EnvironmentPolicy {
                    approve_role: Role::Developer,
                    ..base
                },
                PolicyError::ApproveRole(Role::Developer),
            ),
            (
                EnvironmentPolicy {
                    approval_ttl_secs: 10,
                    ..base
                },
                PolicyError::ApprovalTtl,
            ),
            (
                EnvironmentPolicy {
                    approval_ttl_secs: MAX_APPROVAL_TTL_SECS + 1,
                    ..base
                },
                PolicyError::ApprovalTtl,
            ),
        ];
        for (policy, error) in cases {
            assert_eq!(policy.validate(), Err(error), "{policy:?}");
        }
    }

    #[test]
    fn restarts_and_handovers_need_no_approval() {
        let p = EnvironmentPolicy {
            required_approvals: 2,
            ..EnvironmentPolicy::production()
        };
        for kind in [
            ChangeKind::Deploy,
            ChangeKind::Rollback,
            ChangeKind::Promotion,
            ChangeKind::Build,
            ChangeKind::Rotation,
        ] {
            assert_eq!(p.approvals_for(kind), 2, "{kind:?}");
        }
        assert_eq!(p.approvals_for(ChangeKind::Restart), 0);
        assert_eq!(p.approvals_for(ChangeKind::Handover), 0);
        assert_eq!(EnvironmentPolicy::open().approvals_for(ChangeKind::Deploy), 0);
    }

    #[test]
    fn roles_are_checked_against_the_policy() {
        let p = EnvironmentPolicy {
            deploy_role: Role::Admin,
            ..EnvironmentPolicy::production()
        };
        assert!(!p.may_deploy(None));
        assert!(!p.may_deploy(Some(Role::Developer)));
        assert!(p.may_deploy(Some(Role::Admin)));
        assert!(p.may_approve(Some(Role::Owner)));
        assert!(!p.may_approve(Some(Role::Developer)));
        assert!(!p.may_approve(Some(Role::Viewer)));
    }

    #[test]
    fn weakening_is_detected_field_by_field() {
        let p = EnvironmentPolicy::production();
        assert!(!p.weakens(&p));
        assert!(EnvironmentPolicy::open().weakens(&p));
        assert!(
            !EnvironmentPolicy {
                required_approvals: 2,
                ..p
            }
            .weakens(&p)
        );
        assert!(
            EnvironmentPolicy {
                approve_role: Role::Admin,
                ..p
            }
            .weakens(&EnvironmentPolicy {
                approve_role: Role::Owner,
                ..p
            })
        );
        assert!(
            EnvironmentPolicy {
                approval_ttl_secs: p.approval_ttl_secs + 1,
                ..p
            }
            .weakens(&p)
        );
        assert!(
            !EnvironmentPolicy {
                deploy_role: Role::Owner,
                ..p
            }
            .weakens(&p)
        );
    }

    #[test]
    fn requesters_never_approve_their_own_change() {
        let run = pending(1, 0);
        assert_eq!(
            decide(&run, "alice", false, Decision::Approve, HASH, 0),
            Err(ApprovalError::SelfApproval)
        );
        assert_eq!(
            decide(&run, "alice", false, Decision::Reject, HASH, 0),
            Err(ApprovalError::SelfApproval),
            "nor reject it: withdrawing is cancellation"
        );
        let by_token = Pending {
            requested_by: "token:alice",
            ..run
        };
        assert_eq!(
            decide(&by_token, "alice", false, Decision::Approve, HASH, 0),
            Err(ApprovalError::SelfApproval),
            "a token acts for its owner"
        );
    }

    #[test]
    fn decisions_need_the_shown_plan_and_an_open_window() {
        let run = pending(1, 0);
        assert_eq!(
            decide(&run, "bob", false, Decision::Approve, &[9], 0),
            Err(ApprovalError::StalePlan)
        );
        assert_eq!(
            decide(&run, "bob", false, Decision::Approve, HASH, 1_000),
            Err(ApprovalError::Expired)
        );
        assert_eq!(
            decide(&run, "bob", true, Decision::Approve, HASH, 0),
            Err(ApprovalError::AlreadyDecided)
        );
        assert_eq!(
            decide(&run, "bob", false, Decision::Approve, HASH, 999),
            Ok(Tally::Approved)
        );
    }

    #[test]
    fn approvals_are_counted_and_one_rejection_cancels() {
        let run = pending(3, 1);
        assert_eq!(
            decide(&run, "bob", false, Decision::Approve, HASH, 0),
            Ok(Tally::Waiting { remaining: 1 })
        );
        assert_eq!(
            decide(&pending(3, 2), "carol", false, Decision::Approve, HASH, 0),
            Ok(Tally::Approved)
        );
        assert_eq!(
            decide(&pending(3, 2), "carol", false, Decision::Reject, HASH, 0),
            Ok(Tally::Rejected)
        );
    }

    #[test]
    fn plan_hashes_round_trip_as_hex() {
        assert_eq!(hex(&[0, 171, 255]), "00abff");
        assert_eq!(unhex("00abff"), Some(vec![0, 171, 255]));
        assert_eq!(unhex("00ABFF"), Some(vec![0, 171, 255]));
        assert_eq!(unhex("abc"), None);
        assert_eq!(unhex("zz"), None);
        assert_eq!(unhex(&"a".repeat(130)), None);
        assert_eq!(principal("user:u1"), "u1");
        assert_eq!(principal("plain"), "plain");
    }

    proptest! {
        /// However many approvers vote, a run is approved exactly when the
        /// required number of distinct non-requesters approved and nobody
        /// rejected first.
        #[test]
        fn a_run_is_approved_only_by_enough_other_people(
            required in 1u8..=MAX_APPROVALS,
            votes in prop::collection::vec((0usize..6, any::<bool>()), 0..20),
        ) {
            let people = ["alice", "bob", "carol", "dave", "erin", "frank"];
            let mut decided = std::collections::BTreeSet::new();
            let mut approved = 0u8;
            let mut outcome = None;
            for (who, approve) in votes {
                if outcome.is_some() {
                    break;
                }
                let run = Pending { approved, ..pending(required, 0) };
                let decision = if approve { Decision::Approve } else { Decision::Reject };
                match decide(&run, people[who], decided.contains(&who), decision, HASH, 0) {
                    Ok(Tally::Waiting { remaining }) => {
                        decided.insert(who);
                        approved += 1;
                        prop_assert_eq!(remaining, required - approved);
                    }
                    Ok(done) => {
                        decided.insert(who);
                        if done == Tally::Approved {
                            approved += 1;
                        }
                        outcome = Some(done);
                    }
                    Err(e) => prop_assert!(matches!(e, ApprovalError::SelfApproval | ApprovalError::AlreadyDecided)),
                }
            }
            prop_assert!(!decided.contains(&0), "alice asked for the change");
            if outcome == Some(Tally::Approved) {
                prop_assert_eq!(approved, required);
            }
            prop_assert!(approved <= required);
        }
    }
}

File: go/hub/internal/core/policy/policy.go
// Package policy has environment policies and deployment approvals (M4.1;
// plan §8.2, §13). It replaces the Rust module kuben-core/src/policy.rs, and
// holds EnvironmentPolicy::for_preview of kuben-core/src/preview.rs.
//
// An environment's policy is an immutable revision: who may start
// deployments there, how many distinct people must approve a change before
// it is delivered, who may approve, and how long a request waits. Protection
// is policy, never inferred from a name. The rules here are pure; the store
// applies them inside the transaction that locks the run.
//
//	planned ──(approvals required)──► awaitingApproval ──(enough approvals)──► pendingDelivery
//	                                       │ reject / expiry
//	                                       ▼
//	                                   cancelled
package policy

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/kerr"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/perm"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/scan"
)

const (
	// MaxApprovals is the most approvals a policy may require.
	MaxApprovals uint8 = 5
	// MinApprovalTTLSecs is the shortest time a change may wait for its
	// approvals.
	MinApprovalTTLSecs uint32 = 5 * 60
	// MaxApprovalTTLSecs is the longest time a change may wait for its
	// approvals.
	MaxApprovalTTLSecs uint32 = 30 * 24 * 3600
	// DefaultApprovalTTLSecs is the default wait for approvals: one week.
	DefaultApprovalTTLSecs uint32 = 7 * 24 * 3600
	// MaxCommentChars is the longest approval comment kept.
	MaxCommentChars = 1024

	// maxHexChars is the longest plan hash read from hex.
	maxHexChars = 128
)

// ErrorKind is the rule a refused policy broke.
type ErrorKind string

// The rules a policy can break.
const (
	KindTooManyApprovals ErrorKind = "tooManyApprovals"
	KindDeployRole       ErrorKind = "deployRole"
	KindApproveRole      ErrorKind = "approveRole"
	KindApprovalTTL      ErrorKind = "approvalTtl"
	KindScanGate         ErrorKind = "scanGate"
)

// Error is why a policy was refused. It is comparable, so callers match it
// with == or errors.As; it is also a validation error to errors.Is.
type Error struct {
	Kind ErrorKind
	// Role is the refused role of a KindDeployRole or KindApproveRole error.
	Role perm.Role
}

func (e Error) Error() string {
	switch e.Kind {
	case KindTooManyApprovals:
		return fmt.Sprintf("at most %d approvals can be required", MaxApprovals)
	case KindDeployRole:
		return fmt.Sprintf("the deploy role `%s` cannot deploy", e.Role)
	case KindApproveRole:
		return fmt.Sprintf("the approve role `%s` cannot approve releases", e.Role)
	case KindApprovalTTL:
		return fmt.Sprintf("approvals must wait between %d and %d seconds", MinApprovalTTLSecs, MaxApprovalTTLSecs)
	case KindScanGate:
		return "the scan gate counts `high` or `critical` findings, with a maximum age of an hour to 90 days"
	}
	return "invalid policy: " + string(e.Kind)
}

// Is makes every policy error match kerr.ErrValidation: the API answers a
// refused policy as a validation failure.
func (e Error) Is(target error) bool { return target == error(kerr.ErrValidation) }

// EnvironmentPolicy is one revision of an environment's policy. The zero
// value is not a valid policy; [Open] is the default one.
type EnvironmentPolicy struct {
	// RequiredApprovals are the distinct approvers a change needs before
	// delivery; 0 for none.
	RequiredApprovals uint8 `json:"requiredApprovals"`
	// DeployRole is the weakest role that may start deployments here.
	DeployRole perm.Role `json:"deployRole"`
	// ApproveRole is the weakest role that may approve deployments here.
	ApproveRole perm.Role `json:"approveRole"`
	// ApprovalTTLSecs is how long a change waits for its approvals before it
	// is cancelled.
	ApprovalTTLSecs uint32 `json:"approvalTtlSecs"`
	// Scan is what vulnerability findings a deployment may carry (M4.6).
	Scan scan.Gate `json:"scan"`
}

// Open is the default policy: no approvals; developers deploy.
func Open() EnvironmentPolicy {
	return EnvironmentPolicy{
		RequiredApprovals: 0,
		DeployRole:        perm.Developer,
		ApproveRole:       perm.Admin,
		ApprovalTTLSecs:   DefaultApprovalTTLSecs,
		Scan:              scan.Off(),
	}
}

// Production is one approval by an admin who did not ask for the change;
// known critical findings are refused.
func Production() EnvironmentPolicy {
	p := Open()
	p.RequiredApprovals = 1
	p.Scan = scan.Production()
	return p
}

// Initial is the first policy of a new environment.
func Initial(production bool) EnvironmentPolicy {
	if production {
		return Production()
	}
	return Open()
}

// ForPreview is a preview's policy, inherited from its source environment's:
// the same roles and vulnerability gate, but no approvals, so every push of
// the pull request deploys.
func ForPreview(source EnvironmentPolicy) EnvironmentPolicy {
	source.RequiredApprovals = 0
	return source
}

// UnmarshalJSON requires every field but `scan`, which defaults to the off
// gate (policies stored before M4.6 have none), and accepts only known roles.
func (p *EnvironmentPolicy) UnmarshalJSON(data []byte) error {
	var w struct {
		RequiredApprovals opt.Val[uint8]     `json:"requiredApprovals"`
		DeployRole        opt.Val[string]    `json:"deployRole"`
		ApproveRole       opt.Val[string]    `json:"approveRole"`
		ApprovalTTLSecs   opt.Val[uint32]    `json:"approvalTtlSecs"`
		Scan              opt.Val[scan.Gate] `json:"scan"`
	}
	if err := json.Unmarshal(data, &w); err != nil {
		return err
	}
	approvals, hasApprovals := w.RequiredApprovals.Get()
	deploy, hasDeploy := w.DeployRole.Get()
	approve, hasApprove := w.ApproveRole.Get()
	ttl, hasTTL := w.ApprovalTTLSecs.Get()
	for _, field := range []struct {
		name    string
		present bool
	}{
		{"requiredApprovals", hasApprovals},
		{"deployRole", hasDeploy},
		{"approveRole", hasApprove},
		{"approvalTtlSecs", hasTTL},
	} {
		if !field.present {
			return kerr.New(kerr.Validation, "missing field `%s`", field.name)
		}
	}
	deployRole, err := perm.ParseRole(deploy)
	if err != nil {
		return fmt.Errorf("deployRole: %w", err)
	}
	approveRole, err := perm.ParseRole(approve)
	if err != nil {
		return fmt.Errorf("approveRole: %w", err)
	}
	*p = EnvironmentPolicy{
		RequiredApprovals: approvals,
		DeployRole:        deployRole,
		ApproveRole:       approveRole,
		ApprovalTTLSecs:   ttl,
		Scan:              w.Scan.Or(scan.Off()),
	}
	return nil
}

// Validate refuses a policy that may not be stored, with an [Error].
func (p EnvironmentPolicy) Validate() error {
	switch {
	case p.RequiredApprovals > MaxApprovals:
		return Error{Kind: KindTooManyApprovals}
	case !p.DeployRole.Grants(perm.AppDeploy):
		return Error{Kind: KindDeployRole, Role: p.DeployRole}
	case !p.ApproveRole.Grants(perm.ReleaseApprove):
		return Error{Kind: KindApproveRole, Role: p.ApproveRole}
	case p.ApprovalTTLSecs < MinApprovalTTLSecs || p.ApprovalTTLSecs > MaxApprovalTTLSecs:
		return Error{Kind: KindApprovalTTL}
	case !p.Scan.Valid():
		return Error{Kind: KindScanGate}
	}
	return nil
}

// ChangeKind is what a deployment run changes, as far as approval is
// concerned. It has no wire form: the store maps its run kinds to it.
type ChangeKind string

// The kinds of change.
const (
	Deploy    ChangeKind = "deploy"
	Rollback  ChangeKind = "rollback"
	Promotion ChangeKind = "promotion"
	// Build is an automatic deploy of a verified build.
	Build ChangeKind = "build"
	// Restart is the same release and configuration, pods replaced.
	Restart ChangeKind = "restart"
	// Handover is the same release and configuration, delivered by the agent.
	Handover ChangeKind = "handover"
	// Rotation is the same release and configuration with a secret's new
	// revision: new values reach production like any other change.
	Rotation ChangeKind = "rotation"
	// Emergency is a person's break-glass rollback (M4.9): it cannot wait
	// for approvals.
	Emergency ChangeKind = "emergency"
)

// ParseChangeKind reads a kind of change.
func ParseChangeKind(s string) (ChangeKind, error) {
	switch k := ChangeKind(s); k {
	case Deploy, Rollback, Promotion, Build, Restart, Handover, Rotation, Emergency:
		return k, nil
	}
	return "", kerr.New(kerr.Validation, "unknown kind of change `%s`", s)
}

// ApprovalsFor is how many approvals a change of kind needs. Restarts and
// handovers change neither the release nor the configuration; anything that
// is not a known kind needs the approvals of a deployment.
func (p EnvironmentPolicy) ApprovalsFor(kind ChangeKind) uint8 {
	switch kind {
	case Restart, Handover, Emergency:
		return 0
	case Deploy, Rollback, Promotion, Build, Rotation:
		return p.RequiredApprovals
	}
	return p.RequiredApprovals
}

// MayDeploy reports whether a caller whose strongest role on the environment
// is role may start a deployment there.
func (p EnvironmentPolicy) MayDeploy(role opt.Val[perm.Role]) bool {
	r, ok := role.Get()
	return ok && r.Rank() >= p.DeployRole.Rank() && r.Grants(perm.AppDeploy)
}

// MayApprove reports whether a caller whose strongest role on the
// environment is role may approve there.
func (p EnvironmentPolicy) MayApprove(role opt.Val[perm.Role]) bool {
	r, ok := role.Get()
	return ok && r.Rank() >= p.ApproveRole.Rank() && r.Grants(perm.ReleaseApprove)
}

// Weakens reports whether p protects less than current in any respect.
// Weakening production protection is an owner's decision.
func (p EnvironmentPolicy) Weakens(current EnvironmentPolicy) bool {
	return p.RequiredApprovals < current.RequiredApprovals ||
		p.DeployRole.Rank() < current.DeployRole.Rank() ||
		p.ApproveRole.Rank() < current.ApproveRole.Rank() ||
		p.ApprovalTTLSecs > current.ApprovalTTLSecs ||
		p.Scan.Weakens(current.Scan)
}

// Decision is an approver's answer. Its wire names are the request's
// (`approve`); [Decision.Stored] is the name of the recorded decision.
type Decision string

// The decisions.
const (
	Approve Decision = "approve"
	Reject  Decision = "reject"
)

// ParseDecision reads a decision.
func ParseDecision(s string) (Decision, error) {
	switch d := Decision(s); d {
	case Approve, Reject:
		return d, nil
	}
	return "", kerr.New(kerr.Validation, "unknown decision `%s`", s)
}

// Stored is the stored name of the decision (`approved`, `rejected`); empty
// for a value that is not a decision.
func (d Decision) Stored() string {
	switch d {
	case Approve:
		return "approved"
	case Reject:
		return "rejected"
	}
	return ""
}

// UnmarshalJSON accepts only the exact wire names.
func (d *Decision) UnmarshalJSON(data []byte) error {
	var text string
	if err := json.Unmarshal(data, &text); err != nil {
		return err
	}
	v, err := ParseDecision(text)
	if err != nil {
		return err
	}
	*d = v
	return nil
}

// ApprovalError is why a decision was refused. Nothing is recorded for a
// refusal. Callers match the constants with == or errors.Is.
type ApprovalError string

// The refusals; the value is the message.
const (
	ErrNotAwaiting    ApprovalError = "the deployment is not waiting for approval"
	ErrExpired        ApprovalError = "the approval window of the deployment has closed"
	ErrSelfApproval   ApprovalError = "whoever asked for a deployment cannot approve it"
	ErrAlreadyDecided ApprovalError = "this approver has already decided on the deployment"
	ErrStalePlan      ApprovalError = "the deployment changed since it was shown: reload and decide again"
)

func (e ApprovalError) Error() string { return string(e) }

// Pending is what a run waiting for approval looks like to a decision.
type Pending struct {
	// RequestedBy is the `kind:principal` of whoever asked for the run.
	RequestedBy string
	Required    uint8
	// Approved are the distinct approvals recorded so far.
	Approved  uint8
	ExpiresAt int64
	PlanHash  []byte
}

// Tally is the outcome of a valid decision.
//
//sumtype:decl
type Tally interface{ tally() }

// Waiting: more approvals are needed.
type Waiting struct{ Remaining uint8 }

// Approved: enough approvals, the run may be delivered.
type Approved struct{}

// Rejected: one rejection cancels the run.
type Rejected struct{}

func (Waiting) tally()  {}
func (Approved) tally() {}
func (Rejected) tally() {}

// Principal is the principal part of `kind:principal`.
func Principal(actor string) string {
	if _, id, found := strings.Cut(actor, ":"); found {
		return id
	}
	return actor
}

// Decide says whether approver (a user id) may record decision on run at
// now, having seen planHash, and what the run becomes. A refusal is an
// [ApprovalError].
func Decide(run Pending, approver string, decidedBefore bool, decision Decision, planHash []byte, now int64) (Tally, error) {
	if now >= run.ExpiresAt {
		return nil, ErrExpired
	}
	// A requester may withdraw nothing and approve nothing: only others decide.
	if Principal(run.RequestedBy) == approver {
		return nil, ErrSelfApproval
	}
	if decidedBefore {
		return nil, ErrAlreadyDecided
	}
	if !bytes.Equal(planHash, run.PlanHash) {
		return nil, ErrStalePlan
	}
	switch decision {
	case Reject:
		return Rejected{}, nil
	case Approve:
		count := run.Approved
		if count < 255 {
			count++
		}
		if count >= run.Required {
			return Approved{}, nil
		}
		return Waiting{Remaining: run.Required - count}, nil
	}
	return nil, kerr.New(kerr.Validation, "unknown decision `%s`", string(decision))
}

// Hex is a plan hash as the API shows it.
func Hex(hash []byte) string { return hex.EncodeToString(hash) }

// Unhex is a plan hash given as hex, in either case; false when it is not
// hex or longer than a plan hash can be.
func Unhex(text string) ([]byte, bool) {
	if len(text)%2 != 0 || len(text) > maxHexChars {
		return nil, false
	}
	hash, err := hex.DecodeString(text)
	if err != nil {
		return nil, false
	}
	return hash, true
}
